### PR TITLE
Added option to keep only the last n checkpoints in PeriodicCheckpointer.

### PR DESCRIPTION
**Pull request by maxfrei750**
## 🚀 Feature
Add option to keep only the most recent n checkpoints in PeriodicCheckpointer.

## Motivation
Checkpoints are quite large and often only the most recent one is of interest.

### CODE REVIEW

File: fvcore/common/checkpoint.py [hunk 1]
```diff
@@ -258,18 +258,26 @@ class PeriodicCheckpointer:
     multiple of period or if `max_iter` is reached.
     """
 
-    def __init__(self, checkpointer: Any, period: int, max_iter: int = None):
+    def __init__(
+        self,
+        checkpointer: Any,
+        period: int,
+        max_iter: int = None,
+        n_keep: int = None,
```
Comment (ppwwyyxx) [inline fvcore/common/checkpoint.py]: Could you rename it to "max_to_keep"? This is how the same argument is called in `tf.train.Saver` & `tf.train.CheckpointManager` and having some consistency is nice. On the design our `Checkpointer` is like `tf.train.Checkpoint` and `PeriodicCheckpointer` is like `tf.train.CheckpointManager`.

File: fvcore/common/checkpoint.py [hunk 2]
```diff
@@ -290,6 +298,15 @@ def step(self, iteration: int, **kwargs: Any):
...
             self.checkpointer.save("model_final", **additional_state)
 
+        if self.n_keep is not None:
+            all_checkpoint_files = self.checkpointer.get_all_checkpoint_files()
+            all_checkpoint_files.sort()
+            files_to_delete = all_checkpoint_files[: -self.n_keep]
+
+            for file in files_to_delete:
+                if os.path.exists(file):
+                    os.remove(file)
```
Comment (ppwwyyxx) [inline fvcore/common/checkpoint.py]: Use `PathManager.exists/rm`.

File: fvcore/common/checkpoint.py [hunk 3]
```diff
@@ -290,6 +298,15 @@ def step(self, iteration: int, **kwargs: Any):
         if iteration >= self.max_iter - 1:
             self.checkpointer.save("model_final", **additional_state)
 
+        if self.n_keep is not None:
```
Comment (ppwwyyxx) [inline fvcore/common/checkpoint.py]: Could you move the whole logic into 
```
if (iteration + 1) % self.period == 0:
```
?

The deletion logic needs to list a directory and therefore potentially slow, based on the underlying storage system. It is sufficient to only run it after saving a new checkpoint, and it's not necessary to worry about an extra "model_final".

File: fvcore/common/checkpoint.py (hunk 1 above)
Comment (maxfrei750) [reply fvcore/common/checkpoint.py]: Done.

File: fvcore/common/checkpoint.py (hunk 2 above)
Comment (maxfrei750) [reply fvcore/common/checkpoint.py]: Done.

File: fvcore/common/checkpoint.py (hunk 3 above)
Comment (maxfrei750) [reply fvcore/common/checkpoint.py]: Done.

**Comment by ppwwyyxx** [comment]
Thanks for the update! Could you add a unittest in `tests/test_checkpoint.py`? We expect all features in fvcore to be tested

File: fvcore/common/checkpoint.py [hunk 4]
```diff
@@ -258,18 +258,26 @@ class PeriodicCheckpointer:
...
+        max_to_keep: int = None,
+    ):
         """
         Args:
             checkpointer (Any): the checkpointer object used to save
             checkpoints.
             period (int): the period to save checkpoint.
             max_iter (int): maximum number of iterations. When it is reached,
                 a checkpoint named "model_final" will be saved.
+            max_to_keep (int): maximum number of checkpoints to keep
```
Comment (wanyenlo) [inline fvcore/common/checkpoint.py]: Could you make it more explicit that the oldest checkpoints will be deleted? Thanks!

**Comment by ppwwyyxx** [comment]
could you run ` pip install --progress-bar off flake8 flake8-bugbear flake8-comprehensions isort black==19.3b0` and `./linter.sh`? The current lint failure is in https://circleci.com/gh/facebookresearch/fvcore/145?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by ppwwyyxx** [comment]
Thanks! Please ignore the current CI failure: it will be fixed in #24

**Comment by wangg12** [comment]
If user restarts the training without changing the output folder, the last_checkpoint would be deleted. So I suggest also keep the last_checkpoint by default.
```
last_ckpt_file = self.checkpointer.get_checkpoint_file()
all_checkpoint_files = [item for item in all_checkpoint_files if not item.endswith("model_final.pth") and last_ckpt_file != item]
```

**Comment by maxfrei750** [comment]
@wangg12 Thanks for the hint. However, I think that now that this PR has been closed, it might be best if you start a new PR.

**Pull request closed without merging**